src/main/scala/com/databricks/labs/overwatch/pipeline/BronzeTransforms.scala

```diff
@@ -434,6 +439,103 @@ trait BronzeTransforms extends SparkSessionWrapper {
 
   }
 
+  private def landClusterEvents(clusterIDs:Array[String],
+                                startTime: TimeTypes,
+                                endTime: TimeTypes,
+                                apiEnv: ApiEnv,
+                                tmpClusterEventsSuccessPath:String,
+                                tmpClusterEventsErrorPath:String)={
+    val finalResponseCount = clusterIDs.length
+    var responseCounter = 0
+    var apiResponseArray = Collections.synchronizedList(new util.ArrayList[String]())
+    var apiErrorArray = Collections.synchronizedList(new util.ArrayList[String]())
+    implicit val ec: ExecutionContextExecutor = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(apiEnv.threadPoolSize))
+    //TODO identify the best practice to implement the future.
+    for (i <- clusterIDs.indices) {
+      val jsonQuery = s"""{"cluster_id":"${clusterIDs(i)}","start_time":${startTime.asUnixTimeMilli},"end_time":${endTime.asUnixTimeMilli},"limit":500}"""
+      val future = Future {
+        val apiObj = ApiCallV2(apiEnv, "clusters/events", jsonQuery, tmpClusterEventsSuccessPath).executeMultiThread()
+        synchronized {
+          apiResponseArray.addAll(apiObj)
+          if (apiResponseArray.size() >= apiEnv.successBatchSize) {
+            Helpers.writeMicroBatchToTempLocation(tmpClusterEventsSuccessPath, apiResponseArray.toString)
+            apiResponseArray = Collections.synchronizedList(new util.ArrayList[String]())
+          }
+        }
+
+      }
+      future.onComplete {
+        case Success(_) =>
+          responseCounter = responseCounter + 1
+
+        case Failure(e) =>
+          if (e.isInstanceOf[ApiCallFailureV2]) {
+            synchronized {
+              apiErrorArray.add(e.getMessage)
+              if (apiErrorArray.size() >= apiEnv.errorBatchSize) {
+                Helpers.writeMicroBatchToTempLocation(tmpClusterEventsErrorPath, apiErrorArray.toString)
+                apiErrorArray = Collections.synchronizedList(new util.ArrayList[String]())
+              }
+            }
+
+          }
+          responseCounter = responseCounter + 1
+          logger.log(Level.ERROR, "Future failure message: " + e.getMessage, e)
+      }
+    }
+    while (responseCounter < finalResponseCount) {
+      //As we are using Futures and running 4 threads in parallel, We are checking if all the treads has completed the execution or not.
+      // If we have not received the response from all the threads then we are waiting for 5 seconds and again revalidating the count.
+      println("Sleeping for 5 seconds...Response received count : " + responseCounter + ",Expected Response count : " + finalResponseCount)
```

Comment:
please change this to logging only unless cumulative sleep time becomes > 120s. There needs to be some timeout or logic to ensure the module cannot get hung forever. Upon timeout the entire module should fail with an appropriate error message that is bubbled up to the error report.

Change the message to `Waiting for other queued API Calls to complete; cumulative wait time <n> seconds`

Reply:
done